misc/mdns/src/dns.rs

```diff
@@ -0,0 +1,375 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Contains methods that handle the DNS encoding and decoding capabilities not available in the
+//! `dns_parser` library.
+
+use data_encoding;
+use libp2p_core::{Multiaddr, PeerId};
+use rand;
+use std::{cmp, error, fmt, str, time::Duration};
+use {META_QUERY_SERVICE, SERVICE_NAME};
+
+/// Decodes a `<character-string>` (as defined by RFC1035) into a `Vec` of ASCII characters.
+// TODO: better error type?
+pub fn decode_character_string(mut from: &[u8]) -> Result<Vec<u8>, ()> {
```

Comment:
That's not implemented, but we need to turn `\\` into `\` and `\"` into `"`.
Since it's quite unlikely to happen in practice, I'm going to use a `Cow` though.